### PR TITLE
Open JSON files lock-free if there are many

### PR DESCRIPTION
**Pull request by lnkuiper**
Helps especially with the remote reading of many files

### CODE REVIEW

File: extension/json/include/json_scan.hpp [hunk 1]
```diff
@@ -235,6 +235,7 @@ struct JSONScanLocalState {
 	void ThrowObjectSizeError(const idx_t object_size);
 	void ThrowInvalidAtEndError();
 
+	//! Must hold the lock
```
Comment (samansmink) [inline extension/json/include/json_scan.hpp:238]: I think our current approach for requiring a lock in a protected/private method is to pass a `unique_lock` reference as a param. Like we do in `CatalogSet::CreateDefaultEntries(CatalogTransaction transaction, unique_lock<mutex> &lock);`

**Review by samansmink — COMMENTED** [review]
Hey @lnkuiper looks good! I added a small nitpick, also: I'm just wondering what would happen in the following case:
say we have 8 thread and 9 files, of which the last one happens to be bigger than the first 8 by factor 100. Would this now result in the 9th file to be read single-threaded or can we now also parallelize the final file if its ndjson?

**Comment by lnkuiper** [comment]
@samansmink if the 9th file is NDJSON, we will not read it in parallel. This has not changed in this PR, this was already the behavior.

File: extension/json/include/json_scan.hpp (hunk 1 above)
Comment (lnkuiper) [reply extension/json/include/json_scan.hpp:238]: The lock must be held for the code within the function. There are two types of uses for this. One where we grab the lock and call the function right after, and one where we've been holding on to the lock and then need to call the function.

I don't understand what using a `unique_lock` and passing it to the function would achieve here. We then would also need to add a parameter `bool i_already_have_the_lock` and then not lock it if this is `true`, or we would need to unlock and then call the function, which would grab the lock again.

**Comment by lnkuiper** [comment]
I have also been thinking about adding a parameter. Currently, `sample_size` means we sample `n` rows, even if it is across JSON files. However, we might want to limit the number of JSON files sampled in this way. What do you think about adding a parameter, something like `maximum_sample_files`, that limits how many files are sampled? @samansmink @Mytherin 

The current sample size is set to 10 vectors, i.e., 20480 rows. However, sampling happens single-threaded, so this might take a lot of time if there are, e.g., 10k remote single-row JSON files, so it might be useful to limit the number of files opened to 32 or 128 or something. The idea is that it stops sampling when it has either seen 10 vectors or 32 files.

**Comment by Mytherin** [comment]
Sure, makes sense to me.

File: extension/json/include/json_scan.hpp (hunk 1 above)
Comment (samansmink) [reply extension/json/include/json_scan.hpp:238]: I think the idea is that if you should only call a function with the lock, making it impossible to call the function without the lock makes things a bit cleaner. I see that we do both throughout the code for protected/private methods: adding a comment that holding the lock is required and passing a unique_lock or lock_guard as an argument (sometimes in a wrapper class like `ClientContextLock`)

I have no strong opinion here, Im fine with merging it like this

Comment (lnkuiper) [reply extension/json/include/json_scan.hpp:238]: I agree with being safe, but I will leave this as it is. Thanks!

**Comment by lnkuiper** [comment]
Thanks for the feedback! I've added the `maximum_sample_files` parameter, which defaults to 32. This needs to be documented.

Once CI passes this is ready to go

**Comment by lnkuiper** [comment]
As discussed today, I've also changed `read_json` to be exactly the same as `read_json_auto` and `read_ndjson` to be the same as `read_ndjson_auto`, i.e., auto-detection is enabled by default, like it is for the CSV reader. This should also be documented

**Pull request merged**